### PR TITLE
Preserve Bundle configuration during app generation

### PR DESCRIPTION
**Pull request by marcotc**
### Summary

When generating a new rails application (`rails new`) using a custom template that 
includes gems from an authenticated source, the user has to provide credentials to 
`bundler`.

One way to do this is by [exporting environment variables](https://bundler.io/v1.16/bundle_config.html#CREDENTIALS-FOR-GEM-SOURCES), for example:

`export BUNDLE_GITHUB__COM=user:pass`: provides credentials for bundler to fetch
gems from `github.com`.

The problem this PR addresses is that we are currently scrubs all `/BUNDLE_.*/`
environment variables by wrapping our system calls in [`Bundler.with_clean_env`](https://github.com/rails/rails/blob/5b5319413a611c949cb796351ff42fa0c0299252/railties/lib/rails/generators/app_base.rb#L381-L388).

We do this because we don't want our commands executed against the generated project
to use the generator's `bundler` environment (e.g. our gems): the generated project should
use it's own configuration.

The problem with `Bundler.with_clean_env` is that, on top of restoring environment
variables to their original state, it also scrubs any `/BUNDLE_.*/` variables, which is harmful for authenticated gem sources.

This PR replaces `Bundler.with_clean_env` with `Bundler.with_original_env`, which only
restores environment variables to their initial state, without additional scrubbing.

### CODE REVIEW

**Comment by y-yagi** [comment]
Related to #22991.

**Comment by y-yagi** [comment]
I tried to reproduce step(described in #6318) with `Bundler.with_original_env`, but not reproduced. I assume this already fixed with bundler side, and no need to use `with_clean_env`.

File: railties/lib/rails/generators/app_base.rb [hunk 1]
```diff
@@ -378,13 +378,17 @@ def bundle_command(command, env = {})
         _bundle_command = Gem.bin_path("bundler", "bundle")
```
Comment (y-yagi) [inline railties/lib/rails/generators/app_base.rb]: Please remove the above `We unset temporary bundler variables to load proper bundler and Gemfile.
` comment.

Comment (marcotc) [reply railties/lib/rails/generators/app_base.rb]: Comment removed.

**Comment by y-yagi** [comment]
@marcotc Thanks!

**Comment by schneems** [comment]
Has this been back-ported to 5.2?

**Comment by y-yagi** [comment]
I did not backport because this could be a breaking change. I confirmed that the issue commented by https://github.com/rails/rails/issues/22991#issuecomment-257810539 fixed with the latest bundler, but I could not find which version solved it.

**Comment by schneems** [comment]
I don't think this is "fixed" in bundler https://github.com/bundler/bundler/issues/7184

**Comment by schneems** [comment]
> I did not backport because this could be a breaking change.

Can you say more here?

I have confirmation that it was "fixed" only by adding the extra method https://github.com/bundler/bundler/issues/7184#issuecomment-498587177.

To get Rails 5.2 to work with docker images that ship with a BUNDLE_PATH either each developer will need to hack their own solution or this needs to be backported (I think).

**Comment by y-yagi** [comment]
Bundler had the issue that not using correct `Gemfile`. This issue could be reproduced with: 

```bash
printf "source :rubygems\n\ngem 'rails'\n" > Gemfile
bundle install
bundle exec rails new app1
rails new app2
ls -l app*/Gemfile.lock 
```  

We used `Bundler.with_clean_env` to work around this problem. #6318, 7fc4b4681b83b86d5871f0726460ea808bd76f1b. 
Since the latest Bundler could not reproduce this issue, So I merged this PR. 

However, it was not possible to determine which bundler solved the issue. Therefore, I am concerned that this issue will occur with the old Ruby + Bundler.

I checked the https://github.com/bundler/bundler/issues/7184, is this not a issue with Spring?  I assume this issue releated that fixed with https://github.com/rails/spring/pull/546.

**Comment by schneems** [comment]
My docker issue happens even when disabling spring.

-- 
Richard Schneeman
https://www.schneems.com

**Comment by y-yagi** [comment]
Thanks for your information! 

I confirmed the issue reproduces without `spring`. OK, the original issue can be solved by updating the bundler(I confirmed that the original issue does not happen also Ruby 2.2 + Bundler 1.17 series), but Docker's issue is difficult to work around.  

I backported this to 5-2-stable b1464e34a1aa2ef6e1c181972dca3a14eb65b291.

**Pull request merged**